fqdn: Fix duplicate CIDRSet generation

tgraf opened the pull request:
When using match patterns, the FQDN cache returns a map of results, each
containing a list of IPs. This can lead to duplicate IPs inserted into the list
of rules.

Fixes: #6541

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/6542)
<!-- Reviewable:end -->

raybejjani commented on pkg/fqdn/helpers.go:
```diff
@@ -52,6 +53,9 @@ func injectToCIDRSetRules(rule *api.Rule, cache *DNSCache, reMap *regexpmap.Rege
 		egressRule := &rule.Egress[egressIdx]
 
+		// Build an IP collection to remove all duplicates
+		allIPs := ip.NewCollection()
```
Could you make this a `[]net.IP` and use `keepUniqueIPs` (https://github.com/cilium/cilium/blob/master/pkg/ip/ip.go#L730) on line 111, where we finally add toCIDRSet rules?

raybejjani reviewed (changes requested):
I'd prefer to use `keepUniqueIPs` (https://github.com/cilium/cilium/blob/master/pkg/ip/ip.go#L730) unless there was another reason you wanted `ip.Collection`. This might have to be backported so it would be a little simpler then.

Edit: I also noticed the unit tests failed. They are brittle, but assume the output list is sorted, which `keepUniqueIPs` keeps.

raybejjani commented on pkg/fqdn/helpers.go:
```diff
@@ -99,10 +103,15 @@ func injectToCIDRSetRules(rule *api.Rule, cache *DNSCache, reMap *regexpmap.Rege
+
+		// Only overwrite ToCIDRSet for rules with FQDN elements
+		if len(allIPs) > 0 {
+			egressRule.ToCIDRSet = api.IPsToCIDRRules(ip.KeepUniqueIPs(allIPs))
```
oh, one last thing. The comment on`injectToCIDRSetRules` already existing rules are preserved. We should remove that line (44) above.

tgraf committed:
fqdn: Fix duplicate CIDRSet generation

When using match patterns, the FQDN cache returns a map of results, each
containing a list of IPs. This can lead to duplicate IPs inserted into the list
of rules.

Fixes: #6541

Signed-off-by: Thomas Graf <thomas@cilium.io>
